### PR TITLE
Downloading cells refactor

### PR DESCRIPTION
**Pull request by nnngrach**
[issue](https://github.com/osmandapp/OsmAnd-Issues/issues/1032)

[video](https://www.dropbox.com/scl/fi/vzwsmnpjgbo2ykadulukg/Screen-Recording-2024-05-09-at-23.10.07.mov?rlkey=xttb7goyvs0corymdmdh4m1m7&dl=0)

### CODE REVIEW

**Comment by tigrim** [comment]
We're writing new functionality in Swift. We need to rewrite the new Obj-C classes added in this PR into Swift!

**Comment by nnngrach** [comment]
video:
https://github.com/osmandapp/OsmAnd-iOS/assets/35684515/f925e225-c42f-44b0-9095-00074cc85c30

File: Sources/Helpers/DownloadingCellHelper/DownloadingCellResourceHelper.swift [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+            progress = task.progressCompleted
+        }
+        
+        if helperHasItemFor(resourceId) {
+            DispatchQueue.main.async { [weak self] in
+
+                self?.setCellProgress(resourceId: resourceId, progress: progress, status: .finished)
+                
+                // Start next downloading if needed
+                if let tasks = OsmAndApp.swiftInstance().downloadsManager.keysOfDownloadTasks(), tasks.count > 0 {
```
Comment (tigrim) [inline Sources/Helpers/DownloadingCellHelper/DownloadingCellResourceHelper.swift:211]: !isEmpty

File: Sources/Helpers/OAResourcesUIHelper.mm [hunk 2]
```diff
@@ -551,6 +557,29 @@ - (BOOL) allItemsDownloaded:(NSArray<OAResourceItem *> *)items
...
 
+- (OAResourceItem *) getActiveItem:(BOOL)useDefautValue
+{
+    for (OAResourceItem *item in _items)
+    {
+        if (item.downloadTask != nil)
+            return item;
+    }
+    if (useDefautValue)
+        return _items[0];
```
Comment (tigrim) [inline Sources/Helpers/OAResourcesUIHelper.mm:568]: check items count

File: Sources/Helpers/OAResourcesUISwiftHelper.mm [hunk 3]
```diff
@@ -36,6 +37,53 @@ - (NSString *) type
     return [OAResourceType resourceTypeLocalized:res.resourceType];
 }
 
+- (EOAOAResourceSwiftItemType) resourceType
+{
+    OAResourceItem *res = (OAResourceItem *)self.objcResourceItem;
+    if (res.resourceType == OsmAndResourceType::Unknown)
```
Comment (tigrim) [inline Sources/Helpers/OAResourcesUISwiftHelper.mm:43]: use  switch (res.resourceType) {

**Review by tigrim — APPROVED** [review]

**Pull request merged**